lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -132,11 +132,6 @@ bool swift::isInstructionTriviallyDead(SILInstruction *inst) {
   if (isa<DebugValueInst>(inst) || isa<DebugValueAddrInst>(inst))
     return false;
 
-  // These invalidate enums so "write" memory, but that is not an essential
-  // operation so we can remove these if they are trivially dead.
-  if (isa<UncheckedTakeEnumDataAddrInst>(inst))
```

Comment:
I would add a special case here for Optional. UncheckedTakeEnumDataAddrInst is only writable for certain multi-payload cases where to extract the payload one needs to "open" the enum. This is not true for optional. Also, even so, I don't understand what the crasher is here. Whats the crash specifically, can you post the error?

Reply:
Here's the error:
```
SIL memory lifetime failure in @unchecked_take_enum_data_addr_not_dead: memory is initialized at function return but shouldn't
memory location: %0 = argument of bb0 : $*FakeOptional<B>
at instruction:   return %1 : $()                                 // id: %2

in function:
// unchecked_take_enum_data_addr_not_dead
sil [ossa] @unchecked_take_enum_data_addr_not_dead : $@convention(thin) (@in FakeOptional<B>) -> () {
bb0(%0 : $*FakeOptional<B>):
  %1 = tuple ()                                   // user: %2
  return %1 : $()                                 // id: %2
} // end sil function 'unchecked_take_enum_data_addr_not_dead'
```

Because `unchecked_take_enum_data_addr` takes (or destroys) the operand. When the instruction is removed, the operand is still initialized.

Reply:
I'll add a special case for `Optional`.

Reply:
Actually, I can't add a special case for optional. It's the same problem (not sure why I thought it wouldn't be at first...) but, it's still an `@in` address. It needs to be destroyed somehow.

Reply:
We could promote it to a `destroy_address` if the uses are empty. But, that's probably better done in sil-combine (or maybe DCE).